mempool/mempool.go

```diff
@@ -654,17 +684,30 @@ func (cache *mapTxCache) Reset() {
 	cache.mtx.Unlock()
 }
 
-// Push adds the given tx to the cache and returns true. It returns false if tx
+// PushTxWithInfo adds the given tx to the cache and returns true. It returns false if tx
 // is already in the cache.
-func (cache *mapTxCache) Push(tx types.Tx) bool {
+func (cache *mapTxCache) PushTxWithInfo(tx types.Tx, txInfo TxInfo) bool {
 	cache.mtx.Lock()
 	defer cache.mtx.Unlock()
 
 	// Use the tx hash in the cache
 	txHash := sha256.Sum256(tx)
 	if moved, exists := cache.map_[txHash]; exists {
 		cache.list.MoveToFront(moved)
-		return false
+		memTx, succ := moved.Value.(*mempoolTx)
+		// succ would be false if tx was reaped
+		if !succ {
+			return false
+		}
+		for i := 0; i < len(memTx.senders); i++ {
+			if txInfo.PeerID == memTx.senders[i] {
+				// TODO: consider punishing peer for dups,
+				// its non-trivial since invalid txs can become valid,
+				// but they can spam the same tx with little cost to them atm.
+				return false
+			}
+		}
```

Comment:
shouldn't we return false here too? The TX was in cache, after all. Otherwise we would have duplicates in the cache because the key is now "hash(tx), peerId", or am I reading this wrong?

Reply:
Good catch! We should return false here just for it to be cleaner. It won't actually cause a problem since we'd then have two linked list entries and just 1 hashmap entry (the key didn't change for the cache), and when going to delete that second linked list entry, the delete call would just be a no-op. (Deleting non-existant hashmap keys doesn't panic)